src/compiler/interpreter.jl

```diff
@@ -173,6 +173,60 @@ end
 
 include("tfunc.jl")
 
+struct EnzymeCache
+    inactive::Bool
+    has_rule::Bool
+end
+
+if VERSION >= v"1.11.0-"
+function CC.ipo_dataflow_analysis!(interp::EnzymeInterpreter, ir::Core.Compiler.IRCode,
+                                   caller::Core.Compiler.InferenceResult)
+    mi = caller.linfo
+    specTypes = simplify_kw(mi.specTypes)
+    inactive = false
+    has_rule = false
+    if is_inactive_from_sig(interp, specTypes, mi)
+        inactive = true
+    else
+        # 2. Check if rule is defined
+        if interp.forward_rules && has_frule_from_sig(interp, specTypes, mi)
+            has_rule = true
+        elseif interp.reverse_rules && has_rrule_from_sig(interp, specTypes, mi)
+            has_rule = true
+        end
+    end
+    CC.stack_analysis_result!(caller, EnzymeCache(inactive, has_rule))
+    @invoke CC.ipo_dataflow_analysis!(interp::Core.Compiler.AbstractInterpreter, ir::Core.Compiler.IRCode,
+                                      caller::Core.Compiler.InferenceResult)
+end
+
+else # v1.10
+# 1.10 doesn't have stack_analysis_result or ipo_dataflow_analysis
+function Core.Compiler.finish(interp::EnzymeInterpreter, opt::Core.Compiler.OptimizationState, ir::Core.Compiler.IRCode,
+                   caller::Core.Compiler.InferenceResult)
+    (; src, linfo) = opt
+    specTypes = simplify_kw(linfo.specTypes)
+    inactive = false
+    has_rule = false
+    if is_inactive_from_sig(interp, specTypes, linfo)
+        inactive = true
+    else
+        # 2. Check if rule is defined
+        if interp.forward_rules && has_frule_from_sig(interp, specTypes, linfo)
+            has_rule = true
+        elseif interp.reverse_rules && has_rrule_from_sig(interp, specTypes, linfo)
+            has_rule = true
+        end
+    end
+    @invoke Core.Compiler.finish(interp::Core.Compiler.AbstractInterpreter, opt::Core.Compiler.OptimizationState,
+                      ir::Core.Compiler.IRCode, caller::Core.Compiler.InferenceResult)
+    # Must happen afterwards
+    if inactive || has_rule
+        Core.Compiler.set_inlineable!(src, false)
+    end
```

Comment:
@aviatesk does this look reasonable to you? 

The goal is to do the rule check once, per method and not once per call,